sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                    }))
+                    .ConfigureAwait(false);
+
+                // Get a complete list of the partition ids present in the Event Hub.  This should be immutable for the time being, but
+                // it may change in the future.
+
+                var partitionIds = await connection.GetPartitionIdsAsync(RetryPolicy).ConfigureAwait(false);
+
+                // Find an ownership to claim and try to claim it.  The method will return null if this instance was not eligible to
+                // increase its ownership list, if no claimable ownership could be found or if a claim attempt has failed.
+
+                var claimedOwnership = await FindAndClaimOwnershipAsync(partitionIds, completeOwnershipList, activeOwnership).ConfigureAwait(false);
+
+                if (claimedOwnership != null)
+                {
+                    InstanceOwnership[claimedOwnership.PartitionId] = claimedOwnership;
+
+                    var context = CreateContext(claimedOwnership.PartitionId);
+                    await InitializeProcessingForPartitionAsync(context).ConfigureAwait(false);
+                }
+
+                // Wait the remaining time, if any, to start the next cycle.  The total time of a cycle defaults to 10 seconds,
+                // but it may be overridden by a derived class.
+
+                TimeSpan remainingTimeUntilNextCycle = LoadBalanceUpdate - cycleDuration.Elapsed;
+
+                if (remainingTimeUntilNextCycle > TimeSpan.Zero)
+                {
+                    // If a stop request has been issued, Task.Delay will throw a TaskCanceledException.  This is expected and it
+                    // will be caught by the StopAsync method.
+
+                    await Task.Delay(remainingTimeUntilNextCycle, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // If cancellation has been requested, throw an exception so we can keep a consistent behavior.
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                throw new TaskCanceledException();
+            }
+        }
+
+        /// <summary>
+        ///   Finds and tries to claim an ownership if this processor instance is eligible to increase its ownership list.
+        /// </summary>
+        ///
+        /// <param name="partitionIds">The set of identifiers for the partitions within the Event Hub that this processor is associated with.</param>
+        /// <param name="completeOwnershipEnumerable">A complete enumerable of ownership obtained from the storage service provided by the user.</param>
+        /// <param name="activeOwnership">The set of ownership that are still active.</param>
+        ///
+        /// <returns>The claimed ownership. <c>null</c> if this instance is not eligible, if no claimable ownership was found or if the claim attempt failed.</returns>
+        ///
+        private Task<PartitionOwnership> FindAndClaimOwnershipAsync(string[] partitionIds,
```

Comment:
You probably want to consider making this `ValueTask` since it seems there's a good chance to complete synchronously.